docs: TEALScript migration guide

joe-p opened the pull request.

joe-p commented on docs/tealscript-migration.md:
```diff
@@ -0,0 +1,508 @@
+necessary outside of other migrations mentioned in this document. More complex arrays might need to use a different class shown below, but
+it is strongly recommended to use fixed-length types whenever possible.
+
+TODO: describe different types of arrays and when you need to use them over native arrays
```
Waiting until all the types have landed so I can get some hands on experience before writing this section

boblat commented on docs/tealscript-migration.md:
```diff
@@ -0,0 +1,587 @@
+
+##### Algorand TypeScript
+
+TODO: How do you get args?
```
we can use `op.arg(n)` to get the arguments.
e.g. https://github.com/algorandfoundation/algorand-typescript-testing/blob/main/examples/htlc-logicsig/signature.algo.ts#L27

boblat commented on docs/tealscript-migration.md:
````diff
@@ -0,0 +1,587 @@
+##### Algorand TypeScript
+
+```ts
+addOne(n: UintN256): UintN256 {
````
We can use `.native` property of `UintN256` to get `bigint` representation. So the example could be written as 
```typescript
addOne(n: UintN256): UintN256 {
  const one = 1n
  // Need to explicitly use UintN256 constructor to get uint256 and use `.native` to perform arithmetic
  const sum = new UintN256(n.native + one)
  return sum
}
```

boblat commented on docs/tealscript-migration.md:
````diff
@@ -0,0 +1,587 @@
+```ts
+addToNumber(n: UintN8) {
+  // Use biguint for intermediate values which can go up to u512
+  const x: biguint = 255
````
we can use `bigint` literal and omit the type hint here. e.g. `const x = 255n`

boblat commented on docs/tealscript-migration.md:
```diff
@@ -0,0 +1,587 @@
+  const x: biguint = 255
+  const sum: biguint = BigUint(n.bytes) + x
+
+  return UintN8(sum - x)
```
we need to call `UintN8` constructor here, e.g. `return new UintN8(sum - x)`

boblat commented on docs/tealscript-migration.md:
````diff
@@ -0,0 +1,587 @@
+##### Algorand TypeScript
+
+```ts
+convertNumber(n: uint64): Uint8N {
````
typo: should be `UintN8`

boblat commented on docs/tealscript-migration.md:
````diff
@@ -0,0 +1,587 @@
+
+```ts
+convertNumber(n: uint64): Uint8N {
+  return UintN8(n)
````
we need to call `UintN8` constructor here, e.g. `return new UintN8(n)`.

boblat commented on docs/tealscript-migration.md:
```diff
@@ -0,0 +1,587 @@
+const b = clone(a)
+b.push(4)
+
+assert(a != b) // b has 4 but a does not
```
`!==` is not supported on `Array<uint64>`. We can assert the length of the arrays, e.g.
```typescript
assertMatch(a, { length: 3 })
assertMatch(b, { length: 4 })
```
or we can compare them against array literals
```typescript
assertMatch(a, [1, 2, 3])
assertMatch(b, [1, 2, 3, 4])
```

boblat commented on docs/tealscript-migration.md:
````diff
@@ -0,0 +1,328 @@
+
+##### TEALScript
+
+```ts
+class DangerousPaymentLsig extends LogicSig {
+  logic(amt: uint64) {
+    assert(this.txn.amount === amt)
+  }
+}
+```
+
+##### Algorand TypeScript
+
+TODO: How do you get args?
+
+```ts
+class DangerousPaymentLsig extends LogicSig {
+  program() {
+    return this.txn.amount === amt
+  }
+}
+```
````
we can use `op.arg(n)` to get the arguments.
e.g. https://github.com/algorandfoundation/algorand-typescript-testing/blob/main/examples/htlc-logicsig/signature.algo.ts#L27

joe-p commented:
@boblat Addressed most of the feedback. I am a bit confused on nested dynamic types though. My understanding was that Pyua will not allow mutability of nested dynamic types when using native arrays/structs. Is that not the case?

boblat commented:
> @boblat Addressed most of the feedback. I am a bit confused on nested dynamic types though. My understanding was that Pyua will not allow mutability of nested dynamic types when using native arrays/structs. Is that not the case?

Since we already supported this for explicit ARC-4 types, and the new native types are basically more flexible variants of these, nested dynamic mutations are supported.

joe-p commented:
Oh nice! So then we don't even need explicit migrations from TEALScript then.
